python/ml4ir/data/tfrecord_reader.py

```diff
@@ -79,72 +88,102 @@ def _parse_sequence_example_fn(sequence_example_proto):
-
-                    def crop_fn():
-                        tf.print(
-                            "\n[WARN] Bad query found. Number of records : ", tf.shape(mask)[1]
-                        )
-                        return image.crop_to_bounding_box(
-                            mask,
-                            offset_height=0,
-                            offset_width=0,
-                            target_height=1,
-                            target_width=max_num_records,
-                        )
-
-                    mask = tf.cond(
-                        tf.shape(mask)[1] < max_num_records,
-                        # Pad if there are missing records
-                        lambda: image.pad_to_bounding_box(
-                            mask,
-                            offset_height=0,
-                            offset_width=0,
-                            target_height=1,
-                            target_width=max_num_records,
-                        ),
-                        # Crop if there are extra records
-                        crop_fn,
-                    )
-                    mask = tf.squeeze(mask)
-
-                    # Check validity of mask
-                    tf.debugging.assert_greater(
-                        tf.cast(tf.reduce_sum(mask), tf.float32), tf.constant(0.0)
-                    )
-
-                    features_dict["mask"] = mask
-
-                feature_tensor = sparse.reset_shape(feature_tensor, new_shape=[1, max_num_records])
+                feature_tensor = sparse.reset_shape(
+                    feature_tensor, new_shape=[1, max_num_records if pad_records else num_records]
+                )
                 feature_tensor = sparse.to_dense(feature_tensor)
-                feature_tensor = tf.squeeze(feature_tensor)
+                feature_tensor = tf.squeeze(feature_tensor, axis=0)
 
-                # If feature is a string, then decode into numbers
-                if feature_layer_info["type"] == FeatureTypeKey.STRING:
-                    feature_tensor = preprocessing.preprocess_text(
-                        feature_tensor, preprocessing_info
-                    )
-            else:
-                raise ValueError("Invalid input : {}".format(feature_name))
+            # If feature is a string, then decode into numbers
+            if feature_layer_info["type"] == FeatureTypeKey.STRING:
+                feature_tensor = preprocessing.preprocess_text(feature_tensor, preprocessing_info)
 
             features_dict[feature_node_name] = feature_tensor
 
         labels = features_dict.pop(feature_config.get_label(key="name"))
 
-        # Check if label is one-hot and correctly masked
-        tf.debugging.assert_equal(tf.cast(tf.reduce_sum(labels), tf.float32), tf.constant(1.0))
+        if not required_only:
+            # Check if label is one-hot and correctly masked
+            tf.debugging.assert_equal(tf.cast(tf.reduce_sum(labels), tf.float32), tf.constant(1.0))
```

Comment:
@jakemannix At training time, I've still retained the assert to check if at least one record was clicked.

Reply:
Yeah, that's fine given how we train now.  In theory, providing a list of *all negative examples* for a query is a valid thing to do, but we'd have to adapt our loss functions to support that.

(In particular, if we did strictly pointwise stuff, with no lists as we have now, there would be no requiremnet that every query had a positive example)

Reply:
Agreed.